code/ship/ship.cpp

```diff
@@ -13630,66 +13630,38 @@ static int primary_out_of_ammo(ship_weapon *swp, int bank)
 	{
 		if (swp->primary_bank_ammo[bank] <= 0)
 		{
-			return 1;
+			return true;
 		}
 	}
 
 	// note: never out of ammo if not ballistic
-	return 0;
-}
-
-static bool ship_has_a_ballistic_primary(const ship *shipp)
-{
-	const ship_weapon *swp = &shipp->weapons;
-
-	for (int i = 0; i < swp->num_primary_banks; i++)
-	{
-		if (Weapon_info[swp->primary_bank_weapons[i]].wi_flags[Weapon::Info_Flags::Ballistic])
-			return true;
-	}
-
 	return false;
 }
 
 /**
  * Return true if a new index gets selected.
  * 
  * @param objp      pointer to object for ship cycling primary
- * @param direction forward == CYCLE_PRIMARY_NEXT, backward == CYCLE_PRIMARY_PREV
+ * @param direction forward == CycleDirection::NEXT, backward == CycleDirection::PREV
  *
  * NOTE: This code can be called for any arbitrary ship.  HUD messages and sounds are only used
  *       for the player ship.
  */
-int ship_select_next_primary(object *objp, int direction)
+bool ship_select_next_primary(object *objp, CycleDirection direction)
 {
-	ship	*shipp;
-	ship_info *sip;
-	ship_weapon *swp;
-	int new_bank;
-	int original_bank;
-	int i;
-
 	Assert(objp != NULL);
 	Assert(objp->type == OBJ_SHIP);
 	Assert(objp->instance >= 0 && objp->instance < MAX_SHIPS);
 
-	shipp = &Ships[objp->instance];
-	sip = &Ship_info[shipp->ship_info_index];
-	swp = &shipp->weapons;
+	auto shipp = &Ships[objp->instance];
+	auto sip = &Ship_info[shipp->ship_info_index];
+	auto swp = &shipp->weapons;
 
-	Assert(direction == CYCLE_PRIMARY_NEXT || direction == CYCLE_PRIMARY_PREV);
+	Assert((swp->current_primary_bank >= 0) && (swp->current_primary_bank < swp->num_primary_banks));
```

Comment:
iirc `current_primary_bank` will be `-1` for ships without primary weapons. Will this function be avoided if they press the cycle primary key in that case?

Reply:
Oh, good catch.  I moved this Assert from its original location; I'll move it back.